erdos/src/dataflow/operators/join.rs

```diff
@@ -1,198 +1,146 @@
-    /// * `output_stream` - Represents an outgoing stream of messages of type D3.
-    pub fn new<F: 'static + Clone + Fn(Vec<D1>, Vec<D2>) -> D3>(
-        config: OperatorConfig<F>,
-        input_stream_left: ReadStream<D1>,
-        input_stream_right: ReadStream<D2>,
-        output_stream: WriteStream<D3>,
-    ) -> Self {
-        let name = match config.name {
-            Some(s) => s,
-            None => format!("JoinOperator {}", config.id),
-        };
-
-        // Package the state with the left stream and add a callback to the new stream.
-        let stateful_stream_left = input_stream_left.add_state(StreamState::<D1>::new());
-        stateful_stream_left.add_callback(Self::on_left_data_callback);
-
-        // Package the state with the right stream and add a callback to the new stream.
-        let stateful_stream_right = input_stream_right.add_state(StreamState::<D2>::new());
-        stateful_stream_right.add_callback(Self::on_right_data_callback);
-
-        let cb = config
-            .arg
-            .unwrap_or_else(|| panic!("{}: no join function provided", name));
-        stateful_stream_left
-            .add_read_stream(&stateful_stream_right)
-            .borrow_mut()
-            .add_write_stream(&output_stream)
-            .borrow_mut()
-            .add_watermark_callback(
-                move |t: &Timestamp,
-                      left_state: &StreamState<D1>,
-                      right_state: &StreamState<D2>,
-                      write_stream: &mut WriteStream<D3>| {
-                    Self::on_watermark_callback(t, left_state, right_state, write_stream, &cb)
-                },
-            );
-
-        Self {
-            phantom_data: PhantomData,
-        }
+impl TimestampJoin {
+    pub fn new() -> Self {
+        Self {}
     }
+}
 
-    /// The function to be called when a message is received on the left input stream.
-    /// This callback adds the data received in the message to the state associated with the
-    /// stream.
-    fn on_left_data_callback(t: &Timestamp, msg: &D1, state: &mut StreamState<D1>) {
-        state.add_msg(t, msg.clone());
+impl<T, U> TwoInOneOut<TimeVersionedState<(Vec<T>, Vec<U>)>, T, U, (T, U)> for TimestampJoin
+where
+    T: Data + for<'a> Deserialize<'a>,
+    U: Data + for<'a> Deserialize<'a>,
+{
+    fn on_left_data(
+        &mut self,
+        ctx: &mut TwoInOneOutContext<TimeVersionedState<(Vec<T>, Vec<U>)>, (T, U)>,
+        data: &T,
+    ) {
+        let timestamp = ctx.get_timestamp().clone();
+        let (left_items, right_items) = ctx.get_current_state().unwrap();
+        left_items.push(data.clone());
+
+        let messages: Vec<_> = right_items
+            .iter()
+            .cloned()
+            .map(|right_item| Message::new_message(timestamp.clone(), (data.clone(), right_item)))
+            .collect();
+        for msg in messages {
+            ctx.get_write_stream().send(msg).unwrap();
+        }
```

Comment:
Do you think we should send messages one by one instead of cloning them all first? My concern is that the downstream operators are blocked until the cloning of all the cartesian products takes place (which for some datatypes might take a long time)

Reply:
Agreed, but unfortunately that doesn't compile naiively because a mutable borrow of the write stream (via the context) conflicts with the mutable borrow of the state (via the context).

I'll try to come up with another solution.